k2/csrc/array_of_ragged.cu

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright      2022  Xiaomi Corporation (authors: Daniel Povey)
+ *                2022  ASLP@NWPU          (authors: Hang Lyu)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "k2/csrc/array_of_ragged.h"
+
+namespace k2 {
+
+ArrayOfRaggedShape::ArrayOfRaggedShape(RaggedShape *srcs, int32_t num_srcs) :
+  num_srcs_(num_srcs) {
+  K2_CHECK_GT(num_srcs, 0);
+  K2_CHECK(srcs);
+
+  // Initialize context and num_axes_.
+  c_ = srcs[0].Context();
+  num_axes_ = srcs[0].NumAxes();
+
+  // Check if they have same num-axes and compatible context.
+  for (int32_t i = 1; i < num_srcs_; ++i) {
+    K2_CHECK_EQ(num_axes_, srcs[i].NumAxes());
+    K2_CHECK(IsCompatible(c_, srcs[i].Context()));
+  }
+
+  // Initialize row_splits, row_ids_ and tot_sizes_.
+  row_splits_ = Array2<int32_t *>(c_, num_axes_ - 1, num_srcs_);
+  row_ids_ = Array2<int32_t *>(c_, num_axes_ - 1, num_srcs_);
+  tot_sizes_ = Array1<int32_t>(c_, num_axes_, 0);
+
+  Array2Accessor<int32_t *> row_splits_acc = row_splits_.Accessor(),
+                            row_ids_acc = row_ids_.Accessor();
+  // Bear in mind, when axis == 0, the TotSize() is row_splits.Dim() - 1.
+  // When 0 < axis < NumAxes(), the TotSize() is row_splits.Back().
+  int32_t tot_sizes_data = tot_sizes_.Data();
+
+  for (int32_t i = 1; i < num_axes_; ++i) {
+    for (int32_t j = 0; j < num_srcs_; ++j) {
+      row_splits_acc(i - 1, j) = srcs[j].RowSplits(i).Data();
```

Comment:
I don't think this line can run on GPUs.

Reply:
We actually need to run this on CPU because the pointers are only available from CPU, but that requires that
we initially have an Array1 on CPU, and then transfer it to GPU.